feat/add package prereleases (#1399)
namehash/ensnode: docs/ensnode.io/src/content/docs/docs/contributing/releases.mdx

import { LinkCard } from '@astrojs/starlight/components';

We use [changesets](https://github.com/changesets/changesets) to manage version bumps and release notes for our monorepo's artifacts (NPM packages, Docker images, and Lambda zips).

## Adding a Changeset to Your PR

When you open a PR with feature or bug fix changes, you'll need to include a changeset file that documents your changes:

1. Run `changeset` or `pnpm changeset` in the repository root.
2. Select the packages/apps your changes affect using the interactive prompt.
3. Choose whether your changes are "major" (breaking changes), "minor" (features), or "patch" (bug fixes).
4. Write a clear description of your changes - this will appear in the release notes.

The `changesets/bot` will automatically comment on your PR to either remind you to add a changeset or confirm the version changes that will happen when your PR is merged.

**Important notes:**
- Changesets are only required for user-facing changes (features, bug fixes).
- You don't need a changeset for documentation changes or internal refactoring.

> - All our packages use "fixed" versioning - they all share the same version number regardless of which package triggered the version bump.

Reviewer: ```suggestion
- All our apps and packages (with an exception as noted below) use "fixed" versioning - they all share the same version number regardless of which app or package triggered the version bump.
- An exception to the above "fixed" versioning is the "fallback-ensapi" app. This is a Lambda containing logic specific to NameHash deployments of ENSNode and is versioned independently.
```

> # Creating a Release

Reviewer: Is there a special reason why we should retain the following workflow after the updates being made in this PR? If there's no special benefit to retaining it, I would prefer to remove it.

`.github/workflows/build_ensnode.yml`

@stevedylandev @shrugs Appreciate your advice.

Reviewer 2: confirming that with the proposed release approach the `build_ensnode.yml` workflow is vestigal. By the way, there's an annoying feature in GitHub, which is that they keep the workflow showing even though it's not in main anymore so that you can access previous history.

Options:
1. rename this to `[Deprecated] Build ENSNode` or something and _then_ remove it, in a followup pr so that the item in this list isn't distracting.
<img width="337" height="445" alt="image" src="https://github.com/user-attachments/assets/3c6d7762-4724-4584-bd9d-e8234bc6def6" />

2. delete the previous runs of that workflow which will remove it from the list https://github.com/orgs/community/discussions/26256#discussioncomment-13849174

i'd suggest option 1 for simplicity, and the perhaps obscure possibility that deleting the workflow would affect artifact availability for those previous versions

Upon the publishing of a new release, your change will be included in the produced artifacts and your contributions will be referenced in the GitHub Release notes. There are three different types of releases that your changes can be included in for produced artifacts.

## Full Release

Workflow File: [release.yml](https://github.com/namehash/ensnode/blob/main/.github/workflows/release.yml)

> If your PR includes a changeset and is merged to `main` then it will automatically be added to a new automated Release PR by the Changesets bot. As more changesets are added to `main` the Release PR will continue to update. Once the Release PR is merged into `main` then it will trigger a stable release with all of the included changesets. It will also create a release on GitHub with autogenerated release notes. When a full release runs it will build and publish all of the monorepo's artifacts (NPM packages, Docker images, and Lambda zips).

Reviewer: ```suggestion
If your PR includes a changeset and is merged to `main` then it will automatically be added to a new automated Release PR by the Changesets bot. As more changesets are added to `main` the Release PR will continue to update. Once a Release PR is merged into `main` it triggers a "full" release that will:

- Build and publish all of the monorepo's artifacts (NPM packages, Docker images, and Lambda zips).
- Create a release on GitHub with autogenerated release notes from all the included changesets.
```

- Preview releases can only be triggered manually.
- Preview releases will include the `@preview-*` tag for published artifacts, followed by either the name of the branch or the custom suffix chosen during the action trigger.
- Published artifacts will advance to the version that was included in the changeset of the selected branch.
- Preview releases can only be triggered by authorized NameHash team members.
- No GitHub releases or tags are created for Preview Releases.

# Selecting a Release for Deployment or Installation

When using ENSNode artifacts, you have several release types to choose from.

## Where to Find Releases

- **NPM Packages**: Available on the [npm registry](https://www.npmjs.com/search?q=%40ensnode) under the `@ensnode` organization
- **Docker Images**: Available on [GitHub Container Registry](https://github.com/orgs/namehash/packages?repo_name=ensnode)
- **GitHub Releases**: Full releases are documented with release notes on the [ENSNode GitHub releases page](https://github.com/namehash/ensnode/releases)

> - **Lambda Zip Artifacts**: Available in the Artifact section of a successful workflow run

Author: @shrugs @lightwalker-eth open to suggestions on a better way to link/reference Lambda zips. As far as I know they can only be seen in successful workflow runs, e.g. https://github.com/namehash/ensnode/actions/runs/20581435373

Reviewer: Suggested actions:

1. Update the docs to help people find these "Actions Artifacts". To help explain this, suggest to link to the docs here: https://github.com/actions/upload-artifact?tab=readme-ov-file#where-does-the-upload-go
2. Update the docs to note how these "Actions Artifacts" are only retained for a max of 90 days, including a link to https://github.com/actions/upload-artifact?tab=readme-ov-file#retention-period for more info

I believe we should update our "Full" release workflow so that it publishes the Lambda .zip in a fully durable way that isn't automatically deleted by GitHub after 90 days.

My understanding of the best practice for this is to include the Lambda .zip as a "Release Artifact" such that our "Full" release workflow would be storing the associated Lambda .zip as an additional file in each GitHub Release as shown here:

<img width="2880" height="1580" alt="CleanShot 2025-12-31 at 12 57 33" src="https://github.com/user-attachments/assets/8d902cc4-ba9c-43aa-82b2-d6ec513125f4" />

It also seems attractive to update our release scripts to automatically publish the associated Lambdas as a .zip file to a specific S3 bucket where it automatically names the .zip file to include details such as:
- Is it associated with a preview, a snapshot, or a full release?
- What is the version number / tag associated with the release?

In this way, whenever we want to update the Lambda that's actually running in AWS, we don't need to worry at all about manually building / uploading to S3. Instead we just update the Terraform scripts to reference the desired pre-published lambda .zip file in the target S3 bucket.

@shrugs @jarmint @stevedylandev advice appreciated on what our next steps should be here. Thanks

## Choosing the Right Release Type

### Pinned Full Release Versions
:::note
The example version below of `1.0.0` is a placeholder. Visit the [Releases Page](https://github.com/namehash/ensnode/releases) to pick a pinned full release.
:::
```bash

> npm install @ensnode/[package-name]@1.0.0

Author: Switched to `[package-name]` and `ensnode/[app-name]` per suggestion from @shrugs

- Preview releases can only be triggered manually.
- Preview releases will include the `@preview-*` tag for published packages/images, followed by either the name of the branch or the custom suffix chosen during the action trigger.
- Published packages/images will advance to the version that was included in the changeset of the selected branch.
- Preview releases can only be triggered by authorized NameHash team members.
- No GitHub releases or tags are created for Preview Releases.

# Selecting a Release for Deployment or Installation

When installing ENSNode NPM packages or Docker images, you have several release types to choose from depending on your needs and risk tolerance.

## Where to Find Releases

- **NPM Packages**: Available on the [npm registry](https://www.npmjs.com/search?q=%40ensnode) under the `@ensnode` organization
- **Docker Images**: Available on [GitHub Container Registry](https://github.com/orgs/namehash/packages?repo_name=ensnode)
- **GitHub Releases**: Full releases are documented with release notes on the [ENSNode GitHub releases page](https://github.com/namehash/ensnode/releases)

## Choosing the Right Release Type

> ### For Production Use: `@latest` (Full Releases)

Reviewer 2: yeah we should encourage people to use a specific version not the `latest` tag itself — we're not abiding by real semver so minors can have breaking changes. if it's trivial, it might even be best to remove the `latest` tag entirely, to avoid anyone doing that?

> ### For Production Use: `@latest` (Full Releases)

Reviewer: @stevedylandev We should also make a high-visibility callout in our releases docs for how ENSNode is not currently using Semantic Versioning and therefore even patch or minor releases may include breaking changes.

> ### For Production Use: `@latest` (Full Releases)

Author: @lightwalker-eth Let me now if this works 67dfba89

- Stable, well-tested releases
- Complete release notes and changelog entries

### For Development/Testing: `@next` (Snapshot releases)
```bash
npm install @ensnode/package-name@next  
docker run ghcr.io/namehash/ensnode:next
```

**When to use:**
- Development environments
- When you want to test upcoming features before they're stable

**What to expect:**
- Latest features and bug fixes from the main branch
- May contain breaking changes or unstable features
- More frequent updates (every commit to main)

### For Feature Testing: `@preview-*` (Preview Releases)
```bash
npm install @ensnode/package-name@preview-branch-name
docker run ghcr.io/namehash/ensnode:preview-branch-name
```

**When to use:**
- Testing specific features before they're merged
- Collaborating on development work
- Validating fixes for specific issues

**What to expect:**
- Experimental and potentially unstable
- Built from active development branches
- May not be compatible with other packages

## Version Selection Best Practices

- Check the GitHub releases page for the latest stable version and release notes
- Pin specific versions in your integrations rather than using tags like `@latest`

> - Use full releases (`@latest`) as much as possible for stability

Reviewer: I see this bullet and the one above. They touch on the idea in my new comment up above but they do not sufficiently bring attention to this key issue.

Upon the publishing of a new release, your change will be included in the produced packages/images and your contributions will be referenced in the GitHub Release notes.

## Preview Releases

> Commits to `main` automatically publish prerelease versions with the `next` dist-tag to NPM. These allow testing of changes before an official release.

Reviewer: ```suggestion
Commits to `main` automatically publish prerelease package versions with the `next` dist-tag to NPM. These enable testing of package changes before an official release.
```

> Commits to `main` automatically publish prerelease versions with the `next` dist-tag to NPM. These allow testing of changes before an official release.

Reviewer: Suggest to document here a link to the specific GitHub Action file that automatically triggers this upon each commit to main.

To manually trigger a preview release for a branch that has not been merged to `main`, follow these steps:

> 1. Navigate to [Actions > Release Preview](https://github.com/namehash/ensnode/actions/workflows/preview-release.yml)

Reviewer: Is this "Release Preview" or "Preview Release"?

Since the context is an action, it seems to me that it should always begin with a verb. Ex: "Release Preview" or "Release Snapshot", etc.

2. Click "Run workflow" and select the target branch
3. Install preview packages with: `npm install @ensnode/package-name@preview-branch-name`

**Important notes:**

> - Preview versions are ephemeral and don't create commits or modify the repository

Reviewer: ```suggestion
- Preview releases are ephemeral and don't create commits or modify the repository
```

### Release Snapshot

> Each commit to `main` will trigger the `snapshot-release.yml` workflow to build and publish packages to NPM and Docker images to the GitHub Container Registry (ghcr). These public releases will be under the tag `@next`, allowing anyone to use the npm packages or docker images associated with each commit to main. To install snapshot releases run `pnpm install @ensnode/package-name@next` or `docker run ghcr.io/namehash/ensnode:next`. If the `@latest` tag is used then it will only install or run the latest stable release.

Reviewer: Please also review and update all references to these files after they are renamed.

> Each commit to `main` will trigger the `snapshot-release.yml` workflow to build and publish packages to NPM and Docker images to the GitHub Container Registry (ghcr). These public releases will be under the tag `@next`, allowing anyone to use the npm packages or docker images associated with each commit to main. To install snapshot releases run `pnpm install @ensnode/package-name@next` or `docker run ghcr.io/namehash/ensnode:next`. If the `@latest` tag is used then it will only install or run the latest stable release.

Reviewer: The last sentence here about the @latest tag feels like it belongs in a different section?

In my mental model, when we write these docs we're writing not only for the creator of releases, but also the consumer of releases.

My question here comes from a desire to support the consumer of releases. Will it be clear for them how to decide what is the right release to use? If you think this is already clear than all good. Just posing the question.

**Important notes:**
- Release snapshots are automatic and cannot be triggered manually.
- Release snapshots are considered a preview of what will come in a stable release, so there may be unforeseen side effects or bugs that have not been addressed yet.

### Release Preview

To manually trigger a preview release for a branch that has not been merged to `main`, follow these steps:

1. Navigate to [Actions > Release Preview](https://github.com/namehash/ensnode/actions/workflows/preview-release.yml)
2. Click "Run workflow" and select from the following options:

>  - Which branch you would like to run the preview release workflow on, ideally the branch for an open PR. The workflow needs a target branch to build and publish for.

Author: Updated to the following: 

```diff
- - Which branch you would like to run the preview release workflow on, ideally the branch for an open PR. The workflow needs a target branch to build and publish for.
+ - Which branch you would like to run the preview release workflow on. You can choose any branch and it will trigger a preview release for it, however it is recommended to use the branch where there is an active PR so the workflow can add a GitHub comment to the PR. The GitHub comment will contain helpful installation info such as the generated release tag, which would otherwise have to be located manually through the action logs or the publishing registries.
```

> Why is it ideal that the selected branch has an open PR?

Updated reason: You can choose any branch, but it's most beneficial to select one with an open PR so the GitHub action can leave a comment with helpful tag information.

> What happens if I select a branch that doesn't have an open PR.

This should not raise an alarm anymore because of the change in language. "You can choose any branch" as it will still publish a preview release of the target branch, but it might be harder to locate the published tag name unless you visit the npm registry or ghcr.

 - Choose whether to build and publish only NPM packages, or to build and publish both NPM packages and Docker images to the GitHub Container Registry (ghcr).

>  - (Optional) Provide a custom suffix for the preview release package tag. For example, if you had a branch called `feat/add-api-route` and left this custom suffix field blank, the preview release would be `@ensnode/package-name@preview-feat-add-api-route`. If you fill in the custom suffix field with `users-route` then the resulting package name would be `@ensnode/package-name@preview-users-route`.

Author: Good catch; it should be tag. Fixed!

**Important notes:**
- Snapshot releases are part of the pre-stable state of the `main` branch and should be installed with caution until a [full release](#full-release) is published.
- Release snapshots are automatic and cannot be triggered manually.
- Snapshot releases will include the `@next` tag for published artifacts.
- Published artifacts will advance to the version that was included in the changeset of the PR merged to `main`.
- The `main` branch of `ensnode` is protected. Only PRs approved by the `ensnode` team can be merged to `main` and trigger a snapshot release.
- No GitHub releases or tags are created for snapshot releases.

## Preview Release

Workflow File: [release_preview.yml](https://github.com/namehash/ensnode/blob/main/.github/workflows/release_preview.yml)

To test or install a package before merging it into `main`, a preview release can be used. To manually trigger a preview release for a branch, follow these steps:

1. Navigate to [Actions > Release Preview](https://github.com/namehash/ensnode/actions/workflows/release_preview.yml)
2. Click "Run workflow" and select from the following options:

>  - The branch on which to run the preview release workflow. Any branch other than `main` can be selected. Using a branch with an active PR allows the workflow to add a GitHub comment containing the installation information. If multiple preview releases are triggered for the same brach/PR, the GitHub comment will update with the latest release.

Reviewer: ```suggestion
 - The branch on which to run the preview release workflow. Any branch other than `main` can be selected. Using a branch with an active PR allows the workflow to add a GitHub comment containing the installation information. If multiple preview releases are triggered for the same branch/PR, the GitHub comment will update with the latest release.
```

>  - The branch on which to run the preview release workflow. Any branch other than `main` can be selected. Using a branch with an active PR allows the workflow to add a GitHub comment containing the installation information. If multiple preview releases are triggered for the same brach/PR, the GitHub comment will update with the latest release.

Reviewer: Is there a special reason for Preview Releases to support the case of a branch without a PR?

That seems like an undesirable source of complexity in people's mental models and in operational processes.

Suggest to:
1. Update the Preview Release workflow such that it aborts if the associated branch does not have a PR associated with it.
2. Simplify the docs for Preview Releases such that all complexity about there maybe not being a PR and therefore maybe not having a PR comment, etc.. is removed. Instead we can write about these ideas with black and white certainty: Each Preview Release is associated with a PR, and the PR will have a PR comment.

Author: Agreed; updated workflow and documentation to require a PR for preview releases. https://github.com/namehash/ensnode/pull/1399/commits/16775e2bfbaca8f38113cac9402fe126f836ebaf

>  - Choose which artifacts to build and publish (NPM packages, Docker images, Lambda zips, or any combination).

Author: Good point. I've updated `release_preview.yml` to include more options for selecting what to publish, as well as updating the docs to make it simplified. 

https://github.com/namehash/ensnode/pull/1399/commits/601872352832db26bcd54f51325928de7da0bcb6

>  - Choose which artifacts to build and publish (NPM packages, Docker images, Lambda zips, or any combination).

Author: [See other comment](https://github.com/namehash/ensnode/pull/1399/files#r2655571593)

- Preview releases will include the `@preview-*` tag for published artifacts, followed by either the name of the branch or the custom suffix chosen during the action trigger.
- Published artifacts will advance to the version that was included in the changeset of the selected branch.
- Preview releases can only be triggered by authorized NameHash team members.
- No GitHub releases or tags are created for Preview Releases.

# Selecting a Release for Deployment or Installation

When using ENSNode artifacts, you have several release types to choose from.

## Where to Find Releases

- **NPM Packages**: Available on the [npm registry](https://www.npmjs.com/search?q=%40ensnode) under the `@ensnode` organization
- **Docker Images**: Available on [GitHub Container Registry](https://github.com/orgs/namehash/packages?repo_name=ensnode)
- **GitHub Releases**: Full releases are documented with release notes on the [ENSNode GitHub releases page](https://github.com/namehash/ensnode/releases)
- **Lambda Zip Artifacts**: Available in the Artifact section of a successful workflow run

## Choosing the Right Release Type

### Pinned Full Release Versions
:::note

> The example version below of `1.0.0` is a placeholder. Visit the [Releases Page](https://github.com/namehash/ensnode/releases) to pick a pinned full release.

Author: Updated

:::
```bash

> npm install @ensnode/[package-name]@1.0.0

Reviewer: For NPM packages, I believe it's fine to use @latest or to remove any explicit version number is probably better.

In theory, the issue with use of "latest" doesn't impact NPM packages. It impacts Docker images.

docker run ghcr.io/namehash/ensnode/[app-name]:1.0.0
```

> Pinned full releases are recommended for those who want to start using ENSNode artifacts. By using a pinned version you can maintain full control over features or patches that might be included. Review the release notes on the [Releases Page](https://github.com/namehash/ensnode/releases) to help decide which version to install.

Author: Thank you both! Updated: 7e778f8a

:::caution

> Never use the `latest` tag for Docker deployments (e.g., `ghcr.io/namehash/ensnode:latest` or `ghcr.io/namehash/ensnode`). The `latest` tag is a mutable pointer that will automatically pull different versions over time, causing unexpected updates and potential breaking changes in your deployment. Always pin to a specific version number from the [GitHub releases page](https://github.com/namehash/ensnode/releases).

Author: In this current PR we are documenting and instructing users not to use the `latest` tag. 

It is possible to completely remove the `latest` tag for future ghcr publishing using the [flavor input](https://github.com/actions-marketplace-validations/docker_metadata-action?tab=readme-ov-file#flavor-input) set to `latest=false`. However this introduces a big problem. Since all of the ENSNode apps already have releases where this option was not set, there is no way to retroactively go back and remove those tags. It means a user who tries to run `docker run ghcr.io/namehash/ensnode/[app-name]:latest` will install the last version that used the `latest` tag.

For example, if we published `ensindexer` with the next patch version `1.3.2` and we have the workflow use `latest=false`, that new version will not have the `latest` tag. However the previous version `1.3.1` will still have the `latest` tag. 

I personally would recommend not taking this path and stick with documentation and instruction that pinned versions should be used instead of the `latest` tag.

:::

### Snapshot Releases
```bash
npm install @ensnode/[package-name]@next  

> docker run ghcr.io/namehash/ensnode/[app-name]:next

Author: Doesn't our framing around snapshot releases help clarify this point? 

> Snapshot releases should be used by those who need to test features or patches before they are included in full releases. These releases follow the `main` branch and are not referenced in the GitHub Releases page. Instead they are installed by using the `next` tag for published artifacts. 

Isn't the big picture is that someone might install a snapshot release to test the behavior on the `main` branch, not necessary build upon it further?

> docker run ghcr.io/namehash/ensnode/[app-name]:next

Reviewer 3: who are primary target users of those snapshot releases? 

**also RE: manual workflow for PR builds:** i'd suggest to make snapshot releases automatic on every commit on already opened PR instead of manual action.

- Published artifacts will advance to the version that was included in the changeset of the selected branch.
- No GitHub releases or tags are created for preview releases.

# Selecting a Release for Deployment or Installation

When using ENSNode artifacts, you have several release types to choose from.

## Where to Find Releases

- **NPM Packages**: Available on the [npm registry](https://www.npmjs.com/search?q=%40ensnode) under the `@ensnode` organization.
- **Docker Images**: Available on [GitHub Container Registry](https://github.com/orgs/namehash/packages?repo_name=ensnode).
- **GitHub Releases**: Full releases are documented with release notes on the [ENSNode GitHub releases page](https://github.com/namehash/ensnode/releases).
- **Lambda Zip Artifacts**: Available in the [Artifact section](https://github.com/actions/upload-artifact?tab=readme-ov-file#where-does-the-upload-go) of a successful workflow run. These Action Artifacts are [retained for 90 days](https://github.com/actions/upload-artifact?tab=readme-ov-file#retention-period).

## Choosing the Right Release Type

### Pinned Full Release Versions
:::note

> In the Docker example below, `[version]` is a placeholder that must be replaced with a specific version number from the [Releases Page](https://github.com/namehash/ensnode/releases). Pinning to specific versions is critical for Docker images because the `latest` tag is a mutable pointer. Each time you pull or redeploy, you may receive a different version, introducing unexpected breaking changes or bugs into your running deployment without warning.

Reviewer: ```suggestion
In the installation examples below, `[version]` is a placeholder that must be replaced with a specific version number from the [Releases Page](https://github.com/namehash/ensnode/releases). Pinning to specific versions is critical for all published assets because the `latest` tag is a mutable pointer. Each time you pull or redeploy, you may receive a different version, introducing unexpected breaking changes or bugs into your running deployment without warning.
```

Goal: See related comment below.

:::

```bash

> npm install @ensnode/[package-name]@latest

Reviewer: I think Matt was correct to flag that we should give the specific version info guidance not only for Docker containers, but for all artifacts including NPM packages.

Therefore we should avoid any docs that reference the use of @latest for anything and replace all with a more abstract idea of "[version]"

docker run ghcr.io/namehash/ensnode/[app-name]:[version]
```

> Pinned full releases are required for those who want to use ENSNode Docker Images. By using a pinned version you can maintain full control over features or patches that might be included. This is specifically important for ENSNode as a new version might require updates to environment variables such as `DATABASE_SCHEMA` or change how ENSNode apps work together. Review the release notes on the [Releases Page](https://github.com/namehash/ensnode/releases) to help decide which version to install.

Reviewer: ```suggestion
Pinned full releases are required for those who want to use any published ENSNode artifacts. By using a pinned version you can maintain full control over features or patches that might be included. This is specifically important for ENSNode as a new version might require updates to environment variables such as `DATABASE_SCHEMA` or change how ENSNode apps work together. Review the release notes on the [Releases Page](https://github.com/namehash/ensnode/releases) to help decide which version to install.
```

Goal: See related comment above.

### Snapshot Releases
:::caution

> The `next` tag is a floating pointer that always references the most recent snapshot release. When using Docker images with the `next` tag, you must run `docker pull` to update your local Docker cache if you want to get the actual latest image. Without pulling, you may continue using an older cached version.

Reviewer: Please see my comment below. I'm concerned about how we're currently describing use of the `next` tag. Advice appreciated.

Reviewer 3: this `:next` tag seems redundant for me and just adds uncertainty. 

the `:latest` tag has it's own use for POCs / Demos / development / test environments / CI builds for pulling cache faster. But as mentioned in already placed document: should never be used for running in production.

Reviewer: @jarmint Let's move this discussion into Slack.

## Choosing the Right Release Type

### For Production Deployments: Pinned Full Release Versions
```bash
npm install @ensnode/package-name@1
docker run ghcr.io/namehash/ensnode:1
```

**When to use:**
- Production deployments and any critical infrastructure
- When you need maximum stability and control over updates

**What to expect:**

> - Stable, well-tested releases with complete release notes

Author: Updated

> - The version will never change - you control when to upgrade

Author: Updated

- No unexpected breaking changes or behavior changes

> **Important:** Never use the `latest` tag for Docker deployments (e.g., `ghcr.io/namehash/ensnode:latest` or `ghcr.io/namehash/ensnode`). The `latest` tag is a mutable pointer that will automatically pull different versions over time, causing unexpected updates and potential breaking changes in your deployment. Always pin to a specific version number from the [GitHub releases page](https://github.com/namehash/ensnode/releases).

Author: To clarify, are you asking to remove any mention of `latest` for Docker images in the docs? Or are you suggesting opening another issue to remove `latest` as a possibility for Docker image publishing?
